lib/jsdom/living/helpers/geometry/matrixDomStringParser.js

```diff
@@ -0,0 +1,397 @@
+  "scaleX",
+  "scaleY",
+  "scaleZ",
+  "rotate3d",
+  "rotate",
+  "rotateX",
+  "rotateY",
+  "rotateZ",
+  "perspective",
+  "skew",
+  "skewX",
+  "skewY"
+];
+// https://www.w3.org/TR/css-values-4/#absolute-length
+const absoluteLengths = ["cm", "mm", "Q", "in", "pc", "pt", "px"];
+const LENGTH_CONVERTERS = {
+  cm(x) {
+    return this.in(x) / 2.54;
+  },
+  mm(x) {
+    return this.cm(x) / 10;
+  },
+  Q(x) {
+    return this.cm(x) / 40;
+  },
+  in(x) {
+    return x * 96;
+  },
+  pc(x) {
+    return this.in(x) / 6;
+  },
+  pt(x) {
+    return this.in(x) / 72;
+  },
+  px(x) {
+    return x;
+  }
+};
+
+// https://drafts.csswg.org/css-values-3/#angles
+const angleUnits = ["deg", "grad", "rad", "turn"];
+const ANGLE_CONVERTERS = {
+  deg(x) {
+    return x;
+  },
+  grad(x) {
+    return this.turn(x) / 400;
+  },
+  rad(x) {
+    return x * (180 / Math.PI);
+  },
+  turn(x) {
+    return x * 360;
+  }
+};
+
+const paramConverter = {
+  _rules: {
+    number(x) {
+      return (SYMBOL.radixPoint.test(x) || !isNaN(x)) && x !== null;
+    },
+    angleUnit(x) {
+      return angleUnits.includes(x);
+    },
+    absoluteLengthUnit(x) {
+      return absoluteLengths.includes(x);
+    }
+  },
+  _validate(rules, param) {
+    rules.forEach(matcher => {
+      if (!matcher(param)) {
+        throw new TypeError(`Failed to validate a parameter, ${rules} ${param}`);
+      }
+    });
+  },
+  _getUnitParam(paramString) {
+    let quantity = "";
+    let unit = "";
+
+    if (typeof paramString !== "string") {
+      throw new TypeError();
+    }
+
+    for (const next of paramString) {
+      if (this._rules.number(next) && unit.length === 0) {
+        quantity += next;
+      } else {
+        unit += next;
+      }
+    }
+
+    if (
+      quantity.length === 0
+    ) {
+      throw new TypeError();
+    }
+
+    return [quantity, unit];
+  },
+  number(param) {
+    const { number } = this._rules;
+    this._validate([number], param);
+
+    return Number(param);
+  },
+  lengthPercentage(param) {
+    const { number, absoluteLengthUnit } = this._rules;
+    const [length, unit] = this._getUnitParam(param);
+    this._validate([number], length);
+    this._validate([absoluteLengthUnit], unit);
+
+    return LENGTH_CONVERTERS[unit](Number(length));
+  },
+  angle(param) {
+    const { number, angleUnit } = this._rules;
+    const [length, unit] = this._getUnitParam(param);
+
+    this._validate([number], length);
+    // All transform functions expect 0 or angle type
+    if (Number(length) !== 0) {
+      this._validate([angleUnit], unit);
+    }
+
+    return ANGLE_CONVERTERS[unit || "deg"](Number(length));
+  }
+};
+
+// https://drafts.csswg.org/css-transforms-1/#two-d-transform-functions
+// https://drafts.csswg.org/css-transforms-2/#three-d-transform-functions
+
+// Contains the definitions for valid parameters per transform function.
+// Preforms conversion of parameters to expected values.
+const transformFnParameterValidation = {
```

Comment:
Could you sort these in a way that's easy to find where each function comes from?

```js
const transformFnParameterValidation = {
  // https://drafts.csswg.org/css-transforms-1/#two-d-transform-functions
  matrix(...params) {},
  translate(...params) {},
  // ...

  // https://drafts.csswg.org/css-transforms-2/#three-d-transform-functions
  matrix3d(...params) {},
  translate3d(...params) {},
  // ...
};
```